File: Sudoku_Solver.py
def display_board(Board):

    for i in range(len(Board)):
        if(i%3 == 0 and i != 0):
            print("-----------------------")
        for j in range(len(Board[0])):
            if(j%3 == 0 and j != 0):
                print("|", end = " ")
            if(j != (len(Board[0]) - 1)):
                print(str(Board[i][j]) + " ", end = "")
            else:
                print(str(Board[i][j]))

def find_empty(Board):

    for i in range(len(Board)):
        for j in range(len(Board[0])):
            if(Board[i][j] == 0):
                return (i, j) # (row, column)
    
    return None

def valid(Board, num, pos):

    # Check Row
    for i in range(len(Board[0])):
        if(Board[pos[0]][i] == num and pos[1] != i):
            return False
    
    # Check Column
    for i in range(len(Board)):
        if(Board[i][pos[1]] == num and pos[0] != i):
            return False

    # Check Box
    Box_X = pos[0] // 3
    Box_Y = pos[1] // 3
    for i in range(Box_X * 3, Box_X * 3 + 3, 1):
        for j in range(Box_Y * 3, Box_Y * 3 + 3, 1):
            if(Board[i][j] == num and pos != (i, j)):
                return False
    
    return True

def solve(Board):

    find = find_empty(Board)

    if(find == None):
        return True
    else:
        x = find[0]
        y = find[1]
    
    for i in range(1, 10, 1):
        if(valid(Board, i, (x, y))):
            Board[x][y] = i

            if solve(Board):
                return True

            Board[x][y] = 0

    return False

rows, columns = list(map(int, input("Enter the value of rows and columns.\n").split()))
Board = []
for i in range(rows):
    Temp = list(map(int, input("Enter the value of " + str(i + 1) + " row.\n").split()))
    Board.append(Temp)
print()
display_board(Board)
print()
solve(Board)
display_board(Board)        

File: Sudoku_Solver.cpp
#include<iostream>
using namespace std;

void display_board(int **Board, int rows, int columns){
    
    for(int i = 0; i < rows; i++)
    {
        if(i%3 == 0 && i != 0)
        {
            cout << "----------------------" << endl;
        }
        for(int j = 0; j < columns; j++)
        {
            if(j%3 == 0 && j != 0)
            {
                cout << "| ";
            }
            if(j != columns - 1)
            {
                cout << Board[i][j] << " ";
            }
            else
            {
                cout << Board[i][j] << endl;
            }
        }
    }
}

int * find_empty(int **Board, int rows, int columns)
{
    int *Pos = (int *)malloc(2*sizeof(int));
    for(int i = 0; i < rows; i++)
    {
        for(int j = 0; j < columns; j++)
        {
            if(Board[i][j] == 0)
            {
                Pos[0] = i;
                Pos[1] = j;
                return Pos;
            }
        }
    }
    return NULL;
}

bool valid(int **Board, int Num, int *Pos, int rows, int columns)
{
    // Check Row
    for(int i = 0; i < columns; i++)
    {
        if(Board[Pos[0]][i] == Num && Pos[1] != i)
        {
            return false;
        }
    }

    // Check Column
    for(int i = 0; i < rows; i++)
    {
        if(Board[i][Pos[1]] == Num && Pos[0] != i)
        {
            return false;
        }
    }

    // Check Box
    int Box_X = Pos[0]/3;
    int Box_Y = Pos[1]/3;
    for(int i = Box_X * 3; i < ((Box_X * 3) + 3); i++)
    {
        for(int j = Box_Y * 3; j < ((Box_Y * 3) + 3); j++)
        {
            if(Board[i][j] == Num && Pos[0] != i && Pos[1] != j)
            {
                return false;
            }
        }
    }

    return true;
}

bool Solve(int ***Board, int rows, int columns)
{
    int *Pos = (int *)malloc(2 * sizeof(int));
    Pos = find_empty(*Board, rows, columns);
    int row, col;
    if(Pos == NULL)
    {
        return true;
    }
    else
    {
        row = Pos[0];
        col = Pos[1];
    }
    
    for(int i = 1; i < 10; i++)
    {
        if(valid(*Board, i, Pos, rows, columns))
        {
            (*Board)[row][col] = i;
            if(Solve(Board, rows, columns))
            {
                return true;
            }
            (*Board)[row][col] = 0;
        }
    }

    return false;
}

int main()
{
    int rows, columns;
    cout << "Enter the value of Rows and Columns." << endl;
    cin >> rows >> columns;
    int **Board = (int **)malloc(rows * sizeof(int *));
    for(int i = 0; i < 9; i++)
    {
        *(Board + i) = (int *)malloc(columns * sizeof(int));
    }
    for(int i = 0; i < rows; i++)
    {
        cout << "Enter the " << (i + 1) << " row of Sudoku" << endl;
        for(int j = 0; j < columns; j++)
        {
            cin >> Board[i][j];  
        }
    }

    display_board(Board, rows, columns);
    cout << endl << endl;
    Solve(&Board, rows, columns);
    display_board(Board, rows, columns);

    return 0;
}
